### PR TITLE
add check-large-weights-bins-on-import and reduce-large-weights-bins-on-import

### PR DESCRIPTION
**Pull request by perrozzi**
add the flags
`check-large-weights-bins-on-import`
`reduce-large-weights-bins-on-import`
to check and remove bins with very large errors, i.e.
`h->GetBinError(i) > h->GetBinContent(i)`

### CODE REVIEW

**Comment by adewit** [comment]
Obviously I didn't follow what was going on today but why are we zeroing bins with large errors? It can easily happen with NLO samples but then that is what it is, either we have to rebin to remove the issue or we have to live with it, but I don't think it can be valid just to ignore the bins (actually if I understand correctly it only sets the bin contents to 0 but keeps the error - so the large error will still enter in the bin-by-bin uncertainty, right?) 
Was there a specific issue that this is meant to solve? If someone can summarise by e-mail perhaps we can find a less drastic solution?

**Comment by perrozzi** [comment]
@arizzi got feedback from giovanni petrucciani and marco peruzzi, I did a first implementation, of course we can discuss and refine

**Comment by perrozzi** [comment]
added SetBinError = 0 when zero-ing large error weights

File: CombineTools/src/CombineHarvester.cc [hunk 1]
```diff
@@ -19,6 +19,8 @@ CombineHarvester::CombineHarvester() : verbosity_(0), log_(&(std::cout)) {
   // }
   flags_["check-negative-bins-on-import"] = true;
   flags_["zero-negative-bins-on-import"] = false;
+  flags_["check-large-weights-bins-on-import"] = false;
+  flags_["zerp-large-weights-bins-on-import"] = false;
```
Comment (arizzi) [inline CombineTools/src/CombineHarvester.cc]: zerp -> zero ?

File: CombineTools/src/Utilities.cc [hunk 2]
```diff
@@ -353,4 +353,24 @@ void ZeroNegativeBins(TH1 *h) {
...
+      has_large = true;
+    }
+  }
+  return has_large;
+}
+
+void ZeroLargeErrorBins(TH1 *h) {
+  for (int i = 1; i <= h->GetNbinsX(); ++i) {
+    if (h->GetBinContent(i) < h->GetBinError(i)) {
+      h->SetBinContent(i, 0.);
```
Comment (arizzi) [inline CombineTools/src/Utilities.cc]: I would not zero that but rather leave it to nominal value

File: CombineTools/src/Utilities.cc [hunk 3]
```diff
@@ -353,4 +353,24 @@ void ZeroNegativeBins(TH1 *h) {
...
+      h->SetBinError(i, 0.);
```
Comment (arizzi) [inline CombineTools/src/Utilities.cc]: I think they were suggesting to put the error to 99% of the bin content

**Comment by perrozzi** [comment]
should be fixed now

**Comment by adewit** [comment]
Ok I still don't know what the underlying problem is, so I'm assuming here but is there a worry about difficulty in the fit being caused by bin-by-bin uncertainties taking the yield in a given bin to 0? We have a single bbb per bin so as long as the total uncertainty (on the sum of all processes) in that bin is smaller than the total bin content this won't be a problem. If I remember correctly we don't have any bins where the total bin uncertainty is larger than the total bin content, so if that is the worry this should address I'm not sure it will help.

**Comment by arizzi** [comment]
the point is that the total error can be large of the total content due to interplay of negative weights and large weights

**Comment by arizzi** [comment]
anyhow if anyone finds a way to get a positive def covariance matrix... then I'm fine... giovanni at some point said that he has seen but did not remember where some branch for a more accurate hessian calculation taking "sigmas" instead of "small steps"... can anyone look around for that piece of combine code?

**Comment by perrozzi** [comment]
anyway, tests didn't seem to change the zll cov matrix (will ask lucaM to cross-check).
@scooperstein suggested to remove a couple of low s/b bdt bins, will try that too

**Comment by adewit** [comment]
Right, what I was saying is that while this can easily happen for individual processes with negative weights it will usually get compensated by the other processes contributing in that bin, especially away from tails. Maybe I’m wrong but I don’t remember any bins that have a total uncertainty larger than the bin content. My guess is the covariance matrix isn’t positive definite because of strong correlations between some of the parameters. Can try running the fast impacts (with crs unblind so similar options to fitdiagnostics for cr-only-postfit plots), it will report any parameters it had to drop to get a positive definite covariance matrix, could start by freezing those. Note that it calculates the covariance matrix in a slightly different way than the standard Hesse in minuit so it’s not guaranteed to solve the problem. Btw I guess Giovanni is remembering this ‘fast impacts’ branch, but at the moment the covariance matrix it calculates doen’t get written into a roofit result, so it can’t then be used as we would with the FitDiagnostics roofit result.

**Pull request merged**